### PR TITLE
agent: Check for -v /dev:/dev case

### PR DESCRIPTION
**Pull request by amshinde**
Introduce a new variable "BindMountDev" to indicate the case where
/dev is passed as a volume on the docker command line.
This case needs to be handled by the agent.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

### CODE REVIEW

**Comment by amshinde** [comment]
cc @sboeuf

File: pkg/hyperstart/types.go [hunk 1]
```diff
@@ -186,6 +186,7 @@ type Container struct {
 	Process       *Process            `json:"process"`
 	RestartPolicy string              `json:"restartPolicy"`
 	Initialize    bool                `json:"initialize"`
+	BindMountDev  bool                `json:"bindMountDev"`
```
Comment (sboeuf) [inline pkg/hyperstart/types.go]: This should be part of a more generic structure like:
```go
type IgnoredMounts struct {
        Dev bool
        Proc bool
        ...
}
```

Make this structure the way you want, but the point is that I don't want to see `Container` structure growing too much. And this is my feeling if we add a new field every time we have to handle a specific case.

Comment (amshinde) [reply pkg/hyperstart/types.go]: @sboeuf How about 
``` 
type IgnoredMountsInfo struct {
  BindMountDev bool    // Flag which says if -v /dev:/dev is true (just Dev does not make it clear at all to me)
  DevShmSize int   // For handing /dev/shm size in the future, currently hardcoded
}
```
I dont know what we will be doing about proc, or if we can do anything about it, so will leave that out for now, we can add those as we start handling them.

Comment (sboeuf) [reply pkg/hyperstart/types.go]: Yes this looks good. I was only using `proc` as an example, but I was not expecting you to use it.
The case with `Shm` looks good and illustrates the fact that we can expect several special handling cases like this to come in the future.

Comment (amshinde) [reply pkg/hyperstart/types.go]: @sboeuf I have pushed changes. Have a look.

File: pkg/hyperstart/types.go [hunk 2]
```diff
@@ -173,19 +173,29 @@ type Process struct {
...
+	Fstype            string              `json:"fstype,omitempty"`
+	Image             string              `json:"image"`
+	Addr              string              `json:"addr,omitempty"`
+	Volumes           []*VolumeDescriptor `json:"volumes,omitempty"`
+	Fsmap             []*FsmapDescriptor  `json:"fsmap,omitempty"`
+	Sysctl            map[string]string   `json:"sysctl,omitempty"`
+	Process           *Process            `json:"process"`
+	RestartPolicy     string              `json:"restartPolicy"`
+	Initialize        bool                `json:"initialize"`
+	BindMountDev      bool                `json:"bindMountDev"`
```
Comment (sboeuf) [inline pkg/hyperstart/types.go]: You forgot to remove this.

Comment (amshinde) [reply pkg/hyperstart/types.go]: Oops! Fixed

**Comment by amshinde** [comment]
@sboeuf  Agree. Deducing information about the system/ignored mounts should be done in container.go , now that we know of atleast two cases, especially deducing the shm size to be done in the future. The dev bind mount was pretty trivial enough, so I had left that to the agent.
I'll add the changes.

**Comment by amshinde** [comment]
@sboeuf Changes pushed

File: container.go [hunk 3]
```diff
@@ -99,6 +99,16 @@ func (c *ContainerConfig) valid() bool {
 	return true
 }
 
+// SystemMountsInfo describes additional information for system mounts that the agent
+// needs to handle
+type SystemMountsInfo struct {
+	// Indicates if /dev has been passed as a bind mount for the host /dev
+	BindMountDev bool
+
+	// Size of /dev/shm assigned by docker.
```
Comment (sboeuf) [inline container.go]: No need to mention Docker from this comment. I would say `Size of /dev/shm assigned by the container specification.`

File: container.go [hunk 4]
```diff
@@ -482,6 +494,17 @@ func (c *Container) fetchState(cmd string) (State, error) {
 	return state, nil
 }
 
+func (c *Container) getSystemMountInfo() {
+	// check if /dev needs to be bind mounted from host /dev
+	for _, m := range c.mounts {
+		if m.Source == "/dev" && m.Destination == "/dev" && m.Type == "bind" {
+			c.systemMountsInfo.BindMountDev = true
```
Comment (sboeuf) [inline container.go]: Don't you think we should maybe remove the mount from the list after we identified it as a special case ?

Comment (amshinde) [reply container.go]: Our hyperstart agent anyways ignores /dev in any case, if /dev is passed as a bind mount or not. I think we should leave it to the agent and not do any removal here.

Comment (sboeuf) [reply container.go]: Yes I agree, let's not remove too much info.

File: container.go (hunk 3 above)
Comment (amshinde) [reply container.go]: fixed

**Comment by sboeuf** [comment]
LGTM

**Pull request merged**